File: Array misc./75: Sort Colors.js
// https://leetcode.com/problems/sort-colors/
// Difficulty: Medium
// tags: bucket sort

// Solution
// bucket sort
// O(n) time and O(n) space. Iterate over the elements and maintain a count for the buckets. Iterate over the array again and rewrite the elements in order.

const sortColors = function (nums) {
  const buckets = {
    0: 0,
    1: 0,
    2: 0,
  };
  for (let num of nums) {
    buckets[num]++;
  }
  console.log(buckets);

  for (let i = 0; i < nums.length; i++) {
    if (buckets[0] > 0) {
      nums[i] = 0;
      buckets[0]--;
    } else if (buckets[1] > 0) {
      nums[i] = 1;
      buckets[1]--;
    } else {
      nums[i] = 2;
      buckets[2]--;
    }
  }

  return nums;
};

File: problems/Sorting/75: Sort Colors.py
# https://leetcode.com/problems/sort-colors/description/
# difficulty: medium
# tags: bucket sort

# Problem
# Given an array nums with n objects colored red, white, or blue, sort them in-place so that objects of the same color are adjacent, with the colors in the order red, white, and blue.

# We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.

# You must solve this problem without using the library's sort function.

# Solution, O(n) time and O(1) space, standard bucket sort

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        counts = Counter(nums)
        c_i = 0
        for i in range(len(nums)):
            while not counts[c_i]:
                c_i += 1
            nums[i] = c_i
            counts[c_i] -= 1

